src/leap/mail/imap/account.py

```diff
@@ -199,44 +151,75 @@ def addMailbox(self, name, creation_ts=None):
+                raise imap4.MailboxCollision, repr(name)
+            return mailboxes
+
+        def set_mbox_creation_ts(collection):
+            d = collection.set_mbox_attr("created", creation_ts)
+            d.addCallback(lambda _: collection)
+            return d
 
-        doc = self._soledad.create_doc(mbox)
-        self._load_mailboxes()
-        return bool(doc)
+        d = self.account.list_all_mailbox_names()
+        d.addCallback(check_it_does_not_exist)
+        d.addCallback(lambda _: self.account.add_mailbox(name))
+        d.addCallback(lambda _: self.account.get_collection_by_mailbox(name))
+        d.addCallback(set_mbox_creation_ts)
+        d.addCallback(self._return_mailbox_from_collection)
+        return d
 
     def create(self, pathspec):
         """
         Create a new mailbox from the given hierarchical name.
 
-        :param pathspec: The full hierarchical name of a new mailbox to create.
-                         If any of the inferior hierarchical names to this one
-                         do not exist, they are created as well.
+        :param pathspec:
+            The full hierarchical name of a new mailbox to create.
+            If any of the inferior hierarchical names to this one
+            do not exist, they are created as well.
         :type pathspec: str
 
-        :return: A true value if the creation succeeds.
-        :rtype: bool
+        :return:
+            A deferred that will fire with a true value if the creation
+            succeeds.
+        :rtype: Deferred
 
         :raise MailboxException: Raised if this mailbox cannot be added.
         """
-        # TODO raise MailboxException
-        paths = filter(
-            None,
-            self._parse_mailbox_name(pathspec).split('/'))
+        paths = filter(None, normalize_mailbox(pathspec).split('/'))
+        subs = []
+        sep = '/'
+
+        def pass_on_collision(failure):
+            failure.trap(imap4.MailboxCollision)
+            return True
+
         for accum in range(1, len(paths)):
-            try:
-                self.addMailbox('/'.join(paths[:accum]))
-            except imap4.MailboxCollision:
-                pass
-        try:
-            self.addMailbox('/'.join(paths))
-        except imap4.MailboxCollision:
+            partial_path = sep.join(paths[:accum])
+            d = self.addMailbox(partial_path)
+            d.addErrback(pass_on_collision)
+            subs.append(d)
+
+        def handle_collision(failure):
+            failure.trap(imap4.MailboxCollision)
             if not pathspec.endswith('/'):
-                return False
-        self._load_mailboxes()
-        return True
+                return defer.succeed(False)
+            else:
+                return defer.succeed(True)
+
+        df = self.addMailbox(sep.join(paths))
+        df.addErrback(handle_collision)
+        subs.append(df)
+
+        def all_good(result):
+            return all(result)
```

Comment:
The code might be more readable if all the functions are together without code intercalated on them. It took me more than 5 minutes to realize that in between the functions there was a for that was not a callback function.